File: backend/app/schemas/predict.py
from pydantic import BaseModel, Field

class PatientVitals(BaseModel):
    HighBP: int = Field(..., ge=0, le=1, description="High blood pressure (0 = no, 1 = yes)")
    HighChol: int = Field(..., ge=0, le=1, description="High cholesterol (0 = no, 1 = yes)")
    CholCheck: int = Field(..., ge=0, le=1, description="Cholesterol check in past 5 years (0 = no, 1 = yes)")
    BMI: float = Field(..., gt=0, description="Body Mass Index")
    Smoker: int = Field(..., ge=0, le=1, description="Have smoked at least 100 cigarettes in entire life (0 = no, 1 = yes)")
    Stroke: int = Field(..., ge=0, le=1, description="Ever told you had a stroke (0 = no, 1 = yes)")
    HeartDiseaseorAttack: int = Field(..., ge=0, le=1, description="Coronary heart disease or myocardial infarction (0 = no, 1 = yes)")
    PhysActivity: int = Field(..., ge=0, le=1, description="Physical activity in past 30 days (0 = no, 1 = yes)")
    Fruits: int = Field(..., ge=0, le=1, description="Consume fruit 1 or more times per day (0 = no, 1 = yes)")
    Veggies: int = Field(..., ge=0, le=1, description="Consume vegetables 1 or more times per day (0 = no, 1 = yes)")
    HvyAlcoholConsump: int = Field(..., ge=0, le=1, description="Heavy alcohol consumption (0 = no, 1 = yes)")
    AnyHealthcare: int = Field(..., ge=0, le=1, description="Have any health care coverage (0 = no, 1 = yes)")
    NoDocbcCost: int = Field(..., ge=0, le=1, description="Could not see doctor because of cost in past 12 months (0 = no, 1 = yes)")
    GenHlth: int = Field(..., ge=1, le=5, description="Self-rated general health (1 = excellent, 5 = poor)")
    MentHlth: int = Field(..., ge=0, le=30, description="Number of days of poor mental health in past 30 days")
    PhysHlth: int = Field(..., ge=0, le=30, description="Number of days of poor physical health in past 30 days")
    DiffWalk: int = Field(..., ge=0, le=1, description="Serious difficulty walking or climbing stairs (0 = no, 1 = yes)")
    Sex: int = Field(..., ge=0, le=1, description="Sex (0 = female, 1 = male)")
    Age: int = Field(..., ge=1, le=13, description="13-level age category (1 = 18-24, 13 = 80+)")
    Education: int = Field(..., ge=1, le=6, description="Education level (1 = elementary, 6 = graduate)")
    Income: int = Field(..., ge=1, le=8, description="Income scale (1 = <$10k, 8 = >=$75k)")

class RiskPredictionResponse(BaseModel):
    risk_probability: float = Field(..., ge=0.0, le=1.0, description="Raw risk probability of chronic disease")
    risk_band: str = Field(..., description="Qualitative risk band based on FIS (Low, Medium, High)")
    confidence: float = Field(..., ge=0.0, le=1.0, description="Fuzzy membership confidence score")

File: frontend/src/api/predict.ts
export interface PatientVitals {
  HighBP: number;               // binary flag (0 or 1), high blood pressure
  HighChol: number;             // binary flag (0 or 1), high cholesterol
  CholCheck: number;            // binary flag (0 or 1), cholesterol check in past 5 years
  BMI: number;                  // body mass index, > 0
  Smoker: number;               // binary flag (0 or 1), smoked at least 100 cigarettes
  Stroke: number;               // binary flag (0 or 1), ever had a stroke
  HeartDiseaseorAttack: number; // binary flag (0 or 1), heart disease or attack
  PhysActivity: number;         // binary flag (0 or 1), physical activity in past 30 days
  Fruits: number;               // binary flag (0 or 1), consume fruit daily
  Veggies: number;              // binary flag (0 or 1), consume vegetables daily
  HvyAlcoholConsump: number;    // binary flag (0 or 1), heavy alcohol consumption
  AnyHealthcare: number;        // binary flag (0 or 1), health care coverage
  NoDocbcCost: number;          // binary flag (0 or 1), could not see doctor due to cost
  GenHlth: number;              // general health scale, 1–5
  MentHlth: number;             // number of poor mental health days, 0–30
  PhysHlth: number;             // number of poor physical health days, 0–30
  DiffWalk: number;             // binary flag (0 or 1), difficulty walking
  Sex: number;                  // binary flag (0 = female, 1 = male)
  Age: number;                  // age bracket category, 1–13
  Education: number;            // education category level, 1–6
  Income: number;               // income category level, 1–8
}

export interface RiskPredictionResponse {
  risk_probability: number; // 0.0–1.0, raw ANN output
  risk_band: 'Low' | 'Medium' | 'High' | string; // from the Fuzzy Inference System
  confidence: number;       // 0.0–1.0, fuzzy membership confidence
}

/**
 * Sends patient vitals to the backend API to predict diabetes risk.
 * @param vitals Patient vitals matching the PatientVitals schema.
 * @returns Prediction response or throws an Error with details.
 */
export async function predictDiabetesRisk(vitals: PatientVitals): Promise<RiskPredictionResponse> {
  const response = await fetch('http://127.0.0.1:8000/api/v1/predict', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(vitals),
  });

  if (!response.ok) {
    let errorMessage = `Server error: ${response.status} ${response.statusText}`;
    try {
      const errorData = await response.json();
      if (errorData && errorData.detail) {
        errorMessage = typeof errorData.detail === 'string'
          ? errorData.detail
          : JSON.stringify(errorData.detail);
      }
    } catch {
      // Fallback if parsing JSON fails
    }
    throw new Error(errorMessage);
  }

  return response.json();
}
